### DISCUSSION TITLE
Are sqlite indexes created asynchronously?

### DISCUSSION THREAD

**Discussion by gd87429** [SeaQL/sea-orm#1092] [Q&A]
Sorry, not necessarily sea-orm related, but hoping someone can point me in the right direction.

In migration, I'm creating a table, followed by a unique index on 2 columns. Everything works as it should. However, immediately after migration I'm inserting data which uses `on conflict do nothing` **using that index**.

Everything works correctly on Postgres, but on SQLite it complains that index does not exist. Funny enough when I run it for the second time on SQLite, everything works as it should. Also if I run it on SQLite using tons of logs (which slows down the program a lot), it also works.

Which makes me wonder if code moves on while index is still being created?? Meaning, is it possible that when `CREATE UNIQUE INDEX ...` is run, the result comes back ok after `.await?`, but the query is still being run in the background asynchronously and that's why the next insert does not recognize the unique index? I'm running out of ideas of what else it could be.

Can't find anything under https://www.sqlite.org/lang_createindex.html

**Comment by billy1624** [comment]
Hey @gd87429, this is indeed a weird behaviour. Can you provide a demo on [dbfiddle.uk](https://dbfiddle.uk/vRbqFpk_)?

**Reply by gd87429** [reply]
@billy1624 Thank you for your time.

dbfiddle did not point out any errors when I copied over my sql statements, which gave me an idea of what else to try.

Because my code is supposed to work with pg, mysql & sqlite, I was setting `min_connections = 5` and `max_connections = 100` for all of them. However, looks like this does not work for sqlite (causes errors like above). As soon as I set both of those values to `1`, all the inserts went through correctly, which means that the index was applied serially (before the inserts have started).

So, I can only assume that if there's more than 1 connection to sqlite via sea-orm, then writes are not guaranteed to be executed serially? Is this normal behavior?

**Comment by ikrivosheev** [comment]
@gd87429 hello! Can you explain your pipeline: migrations, queries? Does queries and migrations run concurrently?

**Reply by ikrivosheev** [reply]
My though was: You run queries before migration is complete.

**Reply by gd87429** [reply]
@ikrivosheev Thanks for your time.

The flow is like this:

1. Connect to pg or mysql or sqlite (pretty much like https://www.sea-ql.org/SeaORM/docs/install-and-config/connection/)
2. Create database if does not exist
3. Connect to newly created database
4. Run migration `#1` that creates a table + a unique index on that table
5. Run migration `#2` that inserts lots of data
6. Crash happens in step 5 because unique index from step 4 is not recognized

The problem seems to be that I was specifying in settings for sqlite to have multiple connections (as I was describing in https://github.com/SeaQL/sea-orm/discussions/1092#discussioncomment-3815154). My impression was that sqlite is supposed to have a lock on writes, so multiple connections are possible, but somehow the inserts were getting ahead of the unique index creation.

For now I fixed it by forcing sqlite to have `min_connections=1` and `max_connections=1`.

**Reply by ikrivosheev** [reply]
Are you waiting for the end of the 4 paragraph?

**Reply by gd87429** [reply]
I think so:

`Migrator::up(&db, None).await?;`
`manager.create_table(...).await?;`
`manager.create_index(...).await?;`
etc...

Is this what you mean?
